libs/grout/src/server.ts

```diff
@@ -0,0 +1,176 @@
+ * router. This allows you to easily mount the Grout API within a larger
+ * Express app like so:
+ *
+ *  const api = createApi({ ... methods ... });
+ *  const app = express();
+ *  router = express.Router();
+ *
+ *  // The router must use the text body parser middleware with content type
+ *  // 'application/json'. Make sure you don't add any other body parsers
+ *  // upstream of the Grout router (e.g. with app.use)
+ *  router.use(express.text({ type: 'application/json' }));
+ *
+ *  registerRoutes(api, router);
+ *
+ *  // Here, we mount the Grout API at /api, but you can mount it anywhere.
+ *  app.use('/api', groutRouter);
+ *
+ * All routes will use the POST HTTP method with a JSON body (for RPC input) and
+ * return a JSON response (for RPC output), using Grout's serialization format
+ * (see the serialization module for details). This allows requests to be
+ * human-readable while still supporting richer types than plain JSON.
+ *
+ * Unexpected exceptions will return a 500 status code with a plain JSON
+ * response containing the error mesage. This will cause the client to throw an
+ * exception.
+ *
+ * You should NOT throw exceptions from your RPC methods for known errors (e.g.
+ * invalid input). Instead, use the Result type to return known errors.  This
+ * ensures that consumers of the method will be forced to handle that error case
+ * explicitly.
+ */
+export function registerRoutes(api: AnyApi, router: express.Router): void {
+  if (!router.stack.some((layer) => layer.name === 'textParser')) {
+    throw new GroutError(
+      'Grout requires the router to use the text body parser middleware with content type application/json.' +
+        " Add it like so: app.use(express.text({ type: 'application/json' }))."
+    );
+  }
+
+  for (const [methodName, method] of Object.entries<AnyRpcMethod>(api)) {
+    const path = `/${methodName}`;
+    debug(`Registering route: ${path}`);
+
+    // All routes use the POST method. This doesn't quite follow the traditional
+    // semantics of HTTP, since there may or may not be a side-effect. But it's
+    // better to err on the side of possible side-effects (as opposed to use GET).
+    // We don't try to use the correct HTTP method because that would require
+    // some way to annotate RPC methods with the appropriate method, which is
+    // more complexity for little practical gain.
+    router.post(path, async (request, response) => {
+      try {
+        debug(`Call: ${methodName}(${request.body})`);
+        if (!isString(request.body)) {
+          throw new GroutError(
+            'Request body was parsed as something other than a string.' +
+              ' Make sure you havent added any other body parsers upstream' +
+              ' of the Grout router - e.g. app.use(express.json()).' +
+              ` Body: ${JSON.stringify(request.body)}`
+          );
+        }
+
+        const input = deserialize(request.body);
```

Comment:
Could we accomplish all our validation / transformation needs while still using a JSON body parser and `res.json`? The current approach makes sense to me! Just sanity checking that we really have to use a text body parser and text responses

Reply:
Yes, we could, but it would require an extra pass over the parsed JSON data, which could slow things down (though not sure it's a necessary optimization at this stage - I was trying to be a bit more performance minded than usual since this will be used in a lot of different places). I also just thought it would be nice to use the built in replacing functionality of JSON.stringify.

Do you think there's an advantage to using Express's JSON serialization?

Reply:
Got it! The main advantage that I was seeing was not needing code to prevent the use of `app.use(express.json())` and also not having to be so careful about how we send responses, e.g. being able to use `res.json` instead of:

```ts
response.set('Content-type', 'application/json');
response.status(200).send(jsonResult);
```

But those are both minor concerns

Reply:
Yeah, I can see the point about being able to use `express.json` for the whole app being easier. Since we have a way to detect that and provide a clear error message, I'm not super worried about it for now. But would like to see if it's a pain in practice. Since I've already got it working this way, I'm going to leave it for now, but will keep an eye on it.